src/metabase/api/caching.clj

```diff
@@ -0,0 +1,150 @@
+(ns metabase.api.caching
+  (:require
+   [clojure.walk :as walk]
+   [compojure.core :refer [GET]]
+   [metabase.api.common :as api]
+   [metabase.api.common.validation :as validation]
+   [metabase.db.query :as mdb.query]
+   [metabase.events :as events]
+   [metabase.public-settings.premium-features :as premium-features :refer [defenterprise]]
+   [metabase.util :as u]
+   [metabase.util.i18n :refer [tru]]
+   [metabase.util.malli.schema :as ms]
+   [toucan2.core :as t2]))
+
+;; Data shape
+
+(defn row->config
+  "Transform from how cache config is stored to how it's used/exposed in the API"
+  [row]
+  {:model    (:model row)
+   :model_id (:model_id row)
+   :strategy (assoc (:config row) :type (:strategy row))})
+
+(defn config->row
+  "Transform cache config from API form into db storage from"
+  [{:keys [model model_id strategy]}]
+  {:model    model
+   :model_id model_id
+   :strategy (:type strategy)
+   :config   (dissoc strategy :type)})
+
+(defn- drop-internal-fields
+  "See `metabase-enterprise.caching.strategies/CacheStrategy`"
+  [schema]
+  (walk/prewalk
+   (fn [x]
+     (if (and (vector? x) (= (first x) :map))
+       (into [] (remove #(:internal (meta %))) x)
+       x))
+   schema))
+
+;; TODO: declaring schema as defenterprise will make OpenAPI docs invalid for an EE version
+(defenterprise CacheStrategy
+  "Schema for a caching strategy"
+  metabase-enterprise.caching.strategies
+  []
+  [:and
+   [:map
+    [:type [:enum :nocache :ttl]]]
+   [:multi {:dispatch :type}
+    [:nocache  [:map ;; not closed due to a way it's used in tests for clarity
+                [:type keyword?]]]
+    [:ttl      [:map {:closed true}
+                [:type [:= :ttl]]
+                [:multiplier ms/PositiveInt]
+                [:min_duration_ms ms/IntGreaterThanOrEqualToZero]]]]])
+
+(defn CacheStrategyAPI
+  "Schema for a caching strategy for the API"
+  []
+  (drop-internal-fields (CacheStrategy)))
+
+(def ^:private CachingModel [:enum "root" "database" "dashboard" "question"])
+
+(defn- assert-valid-models [model ids premium?]
+  (cond
+    (= model "root")
+    (when-not (= ids [0])
```

Comment:
Based on the last PR's discussion, I think it's worth changing to `(some zero? ids)` 😄

Reply:
done